db2util.c

```diff
@@ -2,701 +2,353 @@
+  set_env_attrs(henv);
+
   /* connect */
-  rc = SQLAllocHandle( SQL_HANDLE_DBC, henv, &hdbc);
-  if (db2util_check_sql_errors(fmt, hdbc, SQL_HANDLE_DBC,   rc) == SQL_ERROR) {
-    return SQL_ERROR;
-  }
-  rc = SQLConnect( (SQLHDBC)hdbc, 
-         (SQLCHAR *)NULL, (SQLSMALLINT)0,
-         (SQLCHAR *)NULL, (SQLSMALLINT)0,
-         (SQLCHAR *)NULL, (SQLSMALLINT)0);
-  if (db2util_check_sql_errors(fmt, hdbc, SQL_HANDLE_DBC, rc) == SQL_ERROR) {
-    return SQL_ERROR;
-  }
-  rc = SQLSetConnectAttr((SQLHDBC)hdbc, SQL_ATTR_DBC_SYS_NAMING, (SQLPOINTER)&attr, SQL_IS_INTEGER);
-  if (db2util_check_sql_errors(fmt, hdbc, SQL_HANDLE_DBC, rc) == SQL_ERROR) {
-    return SQL_ERROR;
-  }
-  rc = SQLSetConnectAttr((SQLHDBC)hdbc, SQL_ATTR_TXN_ISOLATION, (SQLPOINTER)&attr_isolation, SQL_IS_INTEGER);
-  if (db2util_check_sql_errors(fmt, hdbc, SQL_HANDLE_DBC, rc) == SQL_ERROR) {
-    return SQL_ERROR;
-  }
+  rc = SQLAllocHandle(SQL_HANDLE_DBC, henv, &hdbc);
+  check_error(hdbc, SQL_HANDLE_DBC, rc);
+  
+  rc = SQLConnect(hdbc, NULL, 0, NULL, 0, NULL, 0);
+  check_error(hdbc, SQL_HANDLE_DBC, rc);
+  set_conn_attrs(hdbc);
+
+
   /* statement */
-  rc = SQLAllocHandle(SQL_HANDLE_STMT, (SQLHDBC) hdbc, &hstmt);
-  if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-    return SQL_ERROR;
-  }
+  rc = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
+  check_error(hstmt, SQL_HANDLE_STMT, rc);
+
   /* prepare */
-  rc = SQLPrepare((SQLHSTMT)hstmt, (SQLCHAR*)stmt_str, (SQLINTEGER)SQL_NTS);
-  if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-    return SQL_ERROR;
-  }
+  rc = SQLPrepare(hstmt, (SQLCHAR*)stmt, SQL_NTS);
+  check_error(hstmt, SQL_HANDLE_STMT, rc);
+
   /* number of input parms */
-  rc = SQLNumParams((SQLHSTMT)hstmt, (SQLSMALLINT*)&nParms);
-  if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-    return SQL_ERROR;
+  rc = SQLNumParams(hstmt, &param_count);
+  check_error(hstmt, SQL_HANDLE_STMT, rc);
+
+  if (param_count != argc) {
+    fprintf(stderr, "Invalid parameter count: expected %d got %d\n", param_count, argc);
+    exit(1);
   }
-  if (nParms > 0) {
-    for (i = 0; i < nParms; i++) {
-      rc = SQLDescribeParam((SQLHSTMT)hstmt, (SQLUSMALLINT)(i + 1), 
-             &sql_data_type, &sql_precision, &sql_scale, &sql_nullable);
-      if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-        return SQL_ERROR;
-      }
-      buff_len[i] = SQL_NTS;
-      rc = SQLBindParameter((SQLHSTMT)hstmt, (SQLUSMALLINT)(i + 1),
-             SQL_PARAM_INPUT, SQL_CHAR, sql_data_type,
-             sql_precision, sql_scale, argv[i], 0, &buff_len[i]);
-      if (db2util_check_sql_errors(fmt, hstmt, SQL_HANDLE_STMT, rc) == SQL_ERROR) {
-        return SQL_ERROR;
-      }
-    }
+  
+  SQLINTEGER input_indicator = SQL_NTS;
+  for (int i = 0; i < param_count; i++) {
+    SQLSMALLINT type;
+    SQLUINTEGER precision;
```

Comment:
Looking at the documentation here: 
[https://www.ibm.com/support/knowledgecenter/ssw_ibm_i_73/cli/rzadpfndescprm.htm](https://www.ibm.com/support/knowledgecenter/ssw_ibm_i_73/cli/rzadpfndescprm.htm) 
Should precision be `SQLINTEGER`?

Reply:
Ah, you're right. That was originally in Tony's code that way. Oddly, it doesn't give a warning. I'll fix that.